### DISCUSSION TITLE
Getting datasource bean that is NOT wrapped

### DISCUSSION THREAD

**Discussion by anderskristian** [micronaut-projects/micronaut-core#8885] [Q&A] [answered]
I have been trying to make an application that uses JDBI3.

To do this I need Micronaut bean of type **DataSource**. 

`
datasources:
  default:
    url: ${JDBC_URL:`jdbc:h2:mem:default;DB_CLOSE_DELAY=-1;DB_CLOSE_ON_EXIT=FALSE`}
    username: ${JDBC_USER:sa}
    password: ${JDBC_PASSWORD:""}
    driverClassName: ${JDBC_DRIVER:org.h2.Driver}

`

I have been struggling with that for several days. Even a week. JDBI is so easy to work with. But getting it to work in Micronaut is NOT easy. 

Micronaut can deliver a **DataSource** bean that is "**wrapped**" into transaction. So Micronaut can make transaction '**seamless**' cross controller and repository adapter.

**_There are probably very good use case for the design Micronaut has taken._** 

I sit back with a feeling that Micronaut has changed the protocol of Java's **DataSource**. 
I find this very difficult to understand. 
And Micronaut certainly does not support my use case.

My current solution is to **_implements BeanCreatedEventListener<DataSource>_** in a factory. And when I receive a DataSource of type **HikariDataSource** I grab the instance and serve that from my factory as "the Bean" produced in the factory.

Random I do not get an event. Then I need to create my own **HikariDataSource** and serve that one.

The reason I have to check that DataSource is a **_HikariDataSource_** it that the datasource could be a **wrappedDataSource**. And that datasource is useless in my use-case.

I'm sitting very frustrated back and feeling that I'm using all my time on understanding Micronaut. Most of the time fighting the system.

A similar problem is "transactions"  this @Transactional annotation that I have fightet aginst 6 months ago (and lost)  - but I will save that for another debate. 

## I have tried this

` 
 @Singleton
  @Primary
  DataSource dataSource;

`

## But I get exception:

`

io.micronaut.transaction.exceptions.NoTransactionException: No current transaction present. Consider declaring @Transactional on the surrounding method

`

## That demonstrate my point

Micronaut have changed the meaning of interface **_DataSource_**

.. And that makes me take many days to use Micronaut .. and always end up making very strange solutions.

**Comment by graemerocher** [comment, marked as answer]
Inject `DataSourceResolver` and call `DataSource unwrapped = resolver.resolve(dataSource)`

**Reply by anderskristian** [reply]
Thanks a lot Graeme
I will use that one

I feel that it might have been easier to explain that Micronaut DATA used a **_TransactionalDataSource_**
So that the DataSource protocol was keept "standard"

Best regards
Anders K

**Reply by anderskristian** [reply]
I can also see that DataSourceResolver needs artefact 

```
    <dependency>
      <groupId>io.micronaut.data</groupId>
      <artifactId>micronaut-data-jdbc</artifactId>
    </dependency>
```
My application did not need **micronaut-data-jdbc** until now :-(

So I still think that Micronaut should consider keeping DataSource in the standard meaning.
